profiling/src/lib.rs

```diff
@@ -336,6 +307,92 @@ pub struct RequestLocals {
     pub vm_interrupt_addr: *const AtomicBool,
 }
 
+/// take a sample every X bytes
+const ALLOCATION_PROFILING_INTERVAL: f32 = 1024.0 * 30.0;
+
+pub struct AllocationProfilingStats {
+    /// number of bytes in "this" sampling interval
+    pub sampling_interval: AtomicU64,
+    /// number of bytes until next sample collection
+    pub remaing_bytes: AtomicI64,
+    /// number of allocations since last sample
+    pub allocation_count: AtomicU64,
+}
+
+impl AllocationProfilingStats {
+    pub fn new() -> AllocationProfilingStats {
+        let next_sampling_interval = AllocationProfilingStats::next_sampling_interval();
+        AllocationProfilingStats {
+            sampling_interval: AtomicU64::new(next_sampling_interval),
+            remaing_bytes: AtomicI64::new(next_sampling_interval as i64),
+            allocation_count: AtomicU64::new(0),
+        }
+    }
+
+    pub fn reset(&self) {
+        self.allocation_count.store(0, Ordering::Relaxed);
+        self.sampling_interval
+            .store(
+                AllocationProfilingStats::next_sampling_interval(),
+                Ordering::Relaxed
+            );
+        self.remaing_bytes
+            .store(
+                self.sampling_interval.load(Ordering::Relaxed) as i64,
+                Ordering::Relaxed
+            );
+    }
+
+    fn next_sampling_interval() -> u64 {
+        Poisson::new(ALLOCATION_PROFILING_INTERVAL)
+            .unwrap()
+            .sample(&mut rand::thread_rng()) as u64
+    }
+
+    pub fn track_allocation(&self, len: u64) -> (bool, u64, u64) {
+        let mut remaing_bytes: i64 = self
+            .remaing_bytes
+            .fetch_sub(len as i64, Ordering::Relaxed)
+            - len as i64;
+
+        if remaing_bytes > 0 {
+            return (false, 0, 0)
+        }
+
+        let allocation_count = self
+            .allocation_count
+            .fetch_add(1, Ordering::Relaxed)
+            + 1;
+
+        let sampling_interval = self.sampling_interval.load(Ordering::Relaxed) as i64;
+        let next_sampling_interval = AllocationProfilingStats::next_sampling_interval();
+        let mut samples: f64 = (remaing_bytes * -1 / sampling_interval as i64) as f64;
+        remaing_bytes = remaing_bytes % sampling_interval;
+
+        loop {
+            remaing_bytes += next_sampling_interval as i64;
+            samples += 1.0;
+            if remaing_bytes > 0 {
+                break;
+            }
+        }
```

Comment:
The code above seems to be a little different from how the Go profiler does it. It's actually fairly important to get all the details right, otherwise it doesn't work correctly. I know this because I recently reimplemented the algorithm for a simulator. It's also difficult for me to review the correctness if you diverge too much from the Go implementation.

So I'd highly recommend you to translate this code verbatim. Otherwise you should add an extensive test suite for correctness.

https://github.com/felixge/alloc-prof-sim/blob/main/main.go#L189-L214

Reply:
It is indeed different from the Go profiler. I had shamelessly stolen that part from the native profiler 😉 
https://github.com/DataDog/ddprof/blob/main/src/lib/allocation_tracker.cc#L164-L185

Reply:
Oh. I was not aware that they came up with their own sampling variant. Looking at [their tests](https://github.com/DataDog/ddprof/blob/main/test/allocation_tracker-ut.cc#L28) I'm worried about correctness.

I'll try to add their algorithm to my simulator ([JIRA](https://datadoghq.atlassian.net/browse/PROF-6628)), but probably not until the next sprint.

For your PR I'd still recommend you to follow the Go implementation closely. It's the only one we got that has been heavily scrutinized for correctness AFAIK.